### PR TITLE
[JENKINS-39109] Add a configuration option for what label to use for docker agents

### PR DESCRIPTION
**Pull request by rsandell**
This adds a folder property and a global configuration where the administrator,
or someone configuring a folder, can specify what label expression to use with
a pipeline with `agent docker: "x"` and omitting the label.
The folder property can be specified on any folder in the hierarchy and the closest
parent to the run is consulted first and then up the tree, lastly the global configuration
is consulted for a value.
An extension point is also introduced with specified ordinals for the existing extensions
with ample room to spare in between in case something wants to jump in between.
Lastly a step that returns the value of the extension point that first returns a non null value.

### CODE REVIEW

**Comment by rsandell** [comment]
@reviewbybees

**Comment by abayer** [comment]
ping @stephenc @michaelneale in particular

**Comment by rsandell** [comment]
I have a bit more motivation behind it in the ticket https://issues.jenkins-ci.org/browse/JENKINS-39109

File: src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.StaplerRequest;
+
+/**
+ * The system config.
+ *
+ * For example the system level {@link DockerLabelProvider}.
+ */
+@Extension @Symbol("pmd")
```
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java]: :bug: `pmd` is already a thing, so I don't think we should be using it as a `Symbol`.

File: src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/FolderConfig.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    public String getDockerLabel() {
+        return dockerLabel;
+    }
+
+    @DataBoundSetter
+    public void setDockerLabel(String dockerLabel) {
+        this.dockerLabel = dockerLabel;
+    }
+
+    @Extension @Symbol("pmd")
```
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/FolderConfig.java]: See other `pmd` `Symbol` concern. =)

File: src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java (hunk 1 above)
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java]: How about `pipeline-model`?

**Comment by abayer** [comment]
Looks good other than the `Symbol` concern, but I want to see others' thoughts on the concept too.

File: src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java (hunk 1 above)
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java]: Declarative inception ;) Using the word declarative in a declarative configuration that generates configuration for a declarative plugin.

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/config/GlobalConfig.java]: Touché! Let's see if anyone else has naming suggestions, but given that, I guess `pipeline-model` may be the best bet.

**Comment by stephenc** [comment]
So :bee: on the code... :ant: on us once again adding another plugin implementation of hierarchical configuration which really should be handled by core / a plugin (cc @HRMPW)

**Comment by jglick** [comment]
Merge conflicts already? Life in the fast line here.

**Comment by michaelneale** [comment]
Live fast or die trying? 

For extra points, see if you can have merge conflicts with yourself (working across machines). It's awesome.

**Comment by stephenc** [comment]
Merge conflicts because bobby's changes were made on the pre-ast module split code base

**Comment by abayer** [comment]
Tentative :bee: (i.e., assuming tests pass)

**Comment by abayer** [comment]
re-:bee:ing 'cos I forgot the symbols too. =)

**Comment by rsandell** [comment]
@reviewbybees done

**Pull request merged**